src/Yesod/Page.hs

```diff
@@ -186,35 +188,36 @@ instance FromJSON position => FromJSON (Position position) where
         <> " or an Object with a \"next\" or \"previous\" key"
 
 newtype Limit = Limit { unLimit :: Int }
+  deriving newtype (Eq, Ord, Show, Num)
```

Comment:
I've been thinking a little more about this and I'm just not sure what's best. My Approve stands, but I thought it might be worth talking about anyway.

As far as I can tell, the options are:

1- The way you have it

This is certainly simplest usage (the caller uses a literal) and a clear interface (the docs will indicate the parameter is a `Limit`). However, it always feels icky to use `Num` just to get a literal working, and this is ultimately exposing unsafe construction without just... exposing the unsafe constructor. I think the ergonomics of the literal are actually the only benefit over exporting the `Limit` constructor itself and using that -- which I'd call option 1b.

With this design, calling `withPage 0` or `withPage (-1)`, however odd, will break later when the endpoint is used, intermittently, and in non-obvious ways. "Programmer error"-style problems should be louder than that, IMO.

2- Accept `Int` and call a safe constructor internally

In this option, we break `something :: Int -> Either String Limit` out of `readLimit` and use that to "parse" an `Int` argument internally in `withPage`. If it's invalid, we would `throw` or even `error`. This is still an unsatisfying runtime error, but at least it'll happen on _every_ use of the endpoint and be very loud. Tests should identify the problem immediately and clearly.

And you retain the ergonomics of a literal.

3- Do (2) but with the `withPage` argument being wrapped as `DefaultLimit`, for the docs

I'm basically neutral on this option. I love `newtype`s like this, but they're less about docs (IMO) and more for a function that accepts multiple arguments at that type, risking transposition errors. I didn't scroll to `withPage` just now to confirm, but I don't think that's the case there.

4- Go all-in on types, with (e.g.) `closed` to implement the `>= 0` validation of a literal at compile time

I don't find the option worth the complexity, personally.

I think my personal leaning is for (1b) or (2), but as mentioned I'm totally fine with (1) or (3). WDYT?

Reply:
I think I'm ok with 2. Ideally we could use `Natural` instead of `Int` (again, for documentation reasons), but that also has an unsafe `Num` constructor

Reply:
`Natural` makes sense. No reason to expose a custom type when a normal ecosystem type would suffice.

Reply:
The problem with `Natural` is that if you pass `withPage (-1)`, our smart constructor will inspect that value and throw a pure error rather than stay in `Either`. If it's an `Int`, we don't get an error from construction

Reply:
I can see an argument that if someone intentionally constructs a negative `Natural`, whatever happens after is on them (i.e. programmer error :man_shrugging: )

Reply:
> No reason to expose a custom type when a normal ecosystem type would suffice.

I avoided `Natural` originally because the constraint is not `>= 0` it's `> 0`. So I needed a smart constructor anyway (hence `Limit`). Whether `Limit` should internally hold an `Int` vs `Natural` is debatable (since both have invalid inhabitants), but clients shouldn't know what's going on there -- which is why `Num` makes me nervous either way. To Chris' point, accepting `Int` to `withPage` (is the only way that) allows us to handle the invalid input ourselves.

Reply:
Yep, updated with `Int` argument and validation

Reply:
LGTM, thanks for thinking about this again.